indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java

```diff
@@ -56,33 +59,29 @@ public SupervisorManager(MetadataSupervisorManager metadataSupervisorManager)
 
   public Optional<SupervisorSpec> getSupervisorSpec(String id)
   {
-    return supervisors.get(id) == null
-           ? Optional.<SupervisorSpec>absent()
-           : Optional.fromNullable(supervisors.get(id).rhs);
+    Pair<Supervisor, SupervisorSpec> supervisor = supervisors.get(id);
+    return supervisor == null ? Optional.<SupervisorSpec>absent() : Optional.fromNullable(supervisor.rhs);
   }
 
-  public boolean hasSupervisor(String id)
-  {
-    return supervisors.containsKey(id);
-  }
-
-  public boolean createAndStartSupervisor(SupervisorSpec spec)
+  public boolean createOrUpdateAndStartSupervisor(SupervisorSpec spec)
   {
+    Preconditions.checkState(started, "SupervisorManager not started");
```

Comment:
@drcrallen yeah I was aware of that but thought it was such an edge case that it would never happen in practice. If you think about how this code is going to be called, it's a very low concurrency section of code. I can make the stopping case safe too but in my opinion it is probably overkill.